rust/src/csv/reader.rs

```diff
@@ -95,17 +96,24 @@ fn build_primitive_array<T: ArrowPrimitiveType>(
 ) -> Result<ArrayRef> {
     let mut builder = PrimitiveArrayBuilder::<T>::new(rows.len());
     for row_index in 0..rows.len() {
+        let is_boolean_type = TypeId::of::<T>() == TypeId::of::<BooleanType>();
         match rows[row_index].get(*col_idx) {
-            Some(s) if s.len() > 0 => match s.parse::<T::Native>() {
-                Ok(v) => builder.push(v)?,
-                Err(_) => {
-                    // TODO: we should surface the underlying error here.
-                    return Err(ArrowError::ParseError(format!(
-                        "Error while parsing value {}",
-                        s
-                    )));
+            Some(s) if s.len() > 0 => {
+                let mut ss = s.to_string();
```

Comment:
There might be a more efficient way of getting this to work without creating a mutable String, but I couldn't win against the borrow-checker.

Regarding doing this outside of this function, I thought of mutating `csv::StringRecord` and changing the case of the inner strings; but I couldn't find a way of doing it (I searched [docs.rs](https://docs.rs/csv/1.0.5/csv/struct.StringRecord.html))

Reply:
Seems unfortunately that we need to create a new string. Can we only do the `to_string` for `BooleanType`? maybe something like:
```rust
    for row_index in 0..rows.len() {
        let is_boolean_type = TypeId::of::<T>() == TypeId::of::<BooleanType>();
        match rows[row_index].get(*col_idx) {
            Some(s) if s.len() > 0 => {
                let t = if is_boolean_type {
                    s.to_lowercase().parse::<T::Native>()
                } else {
                    s.parse::<T::Native>()
                };
                match t {
                    Ok(v) => builder.push(v)?,
                    Err(_) => {
                        // TODO: we should surface the underlying error here.
                        return Err(ArrowError::ParseError(format!(
                            "Error while parsing value {}",
                            s
                        )));
                    }
                }
            }
            _ => builder.push_null().unwrap(),
        }
    }
```

Also, we are checking `is_boolean_type` inside the loop. Please move it outside. Could we also make the `build_primitive_array` a member function of `Reader`? and use the schema to check the type?

Reply:
Thanks @sunchao, I moved the `is_boolean_type` check outside the loop, I missed that.

I think that even if I make the `build_primitive_array` a (private?) member function, I can't improve on what it's currently doing (the part about using the schema to check the type).

```rust
fn build_primitive_array<T: ArrowPrimitiveType>(
    rows: &[StringRecord],
    col_idx: &usize,
) -> Result<ArrayRef>;
```

If I change it to take the schema, I won't know which columns the schema maps to, as I could be supplying a projection of 3 columns out of a 5 column csv.

```rust
impl Reader {
    // thinking of it, this won't work, so it's probably not what you mean
    // there's Utf8 arrays which follow a different path, 
    // and date/time will probably not use the same build_primitive_arrays
    fn build_primitive_arrays(
        rows: &[StringRecord],
        schema: Arc<Schema>,
    ) -> Result<Vec<ArrayRef>>;
}
```

Or am I not reasoning about it well? I still have a lot to learn on Rust

Reply:
Thanks for the update.

> If I change it to take the schema, I won't know which columns the schema maps to, as I could be supplying a projection of 3 columns out of a 5 column csv.

With the `col_idx`, you should get the corresponding type by looking up `self.schema`:

```rust
    fn build_primitive_array<T: ArrowPrimitiveType>(
        &self,
        rows: &[StringRecord],
        col_idx: usize,
    ) -> Result<ArrayRef> {
        let mut builder = PrimitiveArrayBuilder::<T>::new(rows.len());
        for row_index in 0..rows.len() {
            let is_boolean_type = *self.schema.field(col_idx).data_type() == DataType::Boolean;
    ....
```

You'll need to update the `projection` to use `into_iter` rather than `iter` - so you can use `usize` instead of `&usize` for `col_idx`.

Reply:
I don't think it'll always work, it only works if the user doesn't supply a projection that excludes some columns.

If I have a csv with 5 cols, and I supply `vec![0,2,4]` as my column projection, and `Schema(bool, int32, utf8)` as my schema for those 3 columns:

```rust
*self.schema.field(col_idx).data_type() == DataType::Boolean;
```

The `col_idx` will refer to 0, 2, and 4 because `csv::StringRecord` will contain all 5 columns of the csv. Indexing the schema's fields could be out of bounds.

What do you think?

Now that I look at it, #3128 is inefficient as I infer the whole schema, ignoring projection. With projection supplied, it might have buggy behaviour (I'll test that now)

Reply:
Hmm. I think `self.schema` is the schema for the whole CSV file, not the projection's.

Reply:
I just came to that realisation now on a different matter 🤔, I was trying to optimise schema inference to ignore columns that aren't projected, but got bounds errors.

I'll follow your approach. Thanks again

Reply:
Done. PTAL